translator/GCP_cloud_function/cloud_function/main.py

```diff
@@ -42,28 +47,73 @@ def translate_newest_icone_to_wzdx(event, context):
+  try:
+    schema_string=open(schema_file_name,'r').read()
+  except FileNotFoundError as e:
+    raise RuntimeError('invalid schema: file does not exist')
+  if not schema_string:
+    raise RuntimeError('invalid schema: empty schema')
+  try:
+    schema_obj= json.loads(schema_string)
+    jsonschema.validate('', schema_obj)  
+  except json.decoder.JSONDecodeError as e :
+    raise RuntimeError('invalid schema: not valid json') from e
+  except jsonschema.SchemaError as e :
+    raise RuntimeError('invalid schema: schema failed validation') from e
+  except jsonschema.ValidationError:
+    return schema_obj
+  return schema_obj
 
-#this function retrieves username and password info from secrets .This will be implemented after correct secrets are implemented.
 def get_ftp_credentials():
+
   secret_client = secretmanager.SecretManagerServiceClient()
-  username_secret_name = os.environ['icone_ftp_username_secret_name']
-  password_secret_name = os.environ['icone_ftp_password_secret_name']
-  project_id = os.environ['project_id']
+  username_secret_name = os.environ.get('icone_ftp_username_secret_name')
+  password_secret_name = os.environ.get('icone_ftp_password_secret_name')
+  project_id = os.environ.get('project_id')
+           
+  if not username_secret_name or not password_secret_name or not project_id:
+    return None
   request = {"name": f"projects/{project_id}/secrets/{username_secret_name}/versions/latest"}
-  response = client.access_secret_version(request)
+  try:
+    response = secret_client.access_secret_version(request)
+  except:
+    logging.error(RuntimeError('unable to get username from secret. Either the secret does not exist or there are no permissions!'))
+    return None, None
   username = response.payload.data.decode("UTF-8")
 
   request = {"name": f"projects/{project_id}/secrets/{password_secret_name}/versions/latest"}
-  response = client.access_secret_version(request)
+  try:
+    response = secret_client.access_secret_version(request)
+  except:
+    logging.error(RuntimeError('unable to get password from secret. Either the secret does not exist or there are no permissions!'))
+    return None, None
   password = response.payload.data.decode("UTF-8")
 
-  return username,password
\ No newline at end of file
+  return username,password
+
+
+def unsupported_messages_callback(message):
+  project_id = os.environ.get('project_id')
+  unsupported_messages_topic_id = os.environ.get('unsupported_messages_topic_id')
+  if not project_id or not unsupported_messages_callback:
+    return False
+  #update this so that this code will execute once, not everytime the function is called
+  publisher = pubsub_v1.PublisherClient()
+  topic_path = publisher.topic_path(project_id, unsupported_messages_topic_id)
+  #publish unsupported messages into pub/sub topic
+  try:
+    future=publisher.publish(topic_path,str.encode(json.dumps(message, indent=2)),origin='auto_icone_translator_ftp cloud function')
```

Comment:
Were we seeing issues with publishing JSON formatted code here? That seems to be the only difference between here and our except line below.

Reply:
What do you think about refactoring to move the `json.dumps` portion out? Something like:

```
def formatMessage(message):
  try:
    msg = json.dumps(message, indent=2)  
  except:
    msg = message  
  return msg
```

Then you could call the new method from the `unsuppoted_messages_callback` function:
```
future = publisher.publish(topic_path, str.encode(formatMessage(message)))
```

Reply:
i updated the unit test according to your instruction

Reply:
Looks good! Can we remove the commented 2nd call to publish?

Reply:
i removed the commented 2nd call